pkg/minikube/machine/cache_images.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machine
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/sync/errgroup"
+
+	"k8s.io/minikube/pkg/minikube/assets"
+	"k8s.io/minikube/pkg/minikube/bootstrapper"
+	"k8s.io/minikube/pkg/minikube/constants"
+
+	"github.com/containers/image/copy"
+	"github.com/containers/image/docker"
+	"github.com/containers/image/docker/archive"
+	"github.com/containers/image/signature"
+	"github.com/containers/image/types"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+)
+
+const tempLoadDir = "/tmp"
+
+func CacheImagesForBootstrapper(clusterBootstrapper string) error {
+	images, ok := bootstrapper.CachedImagesForBootstrapper[clusterBootstrapper]
+	if !ok {
+		glog.Infoln("Could not find list of images to cache for bootstrapper %s", clusterBootstrapper)
+		return nil
+	}
+
+	if err := CacheImages(images, constants.ImageCacheDir); err != nil {
+		return errors.Wrapf(err, "Caching images for %s", clusterBootstrapper)
+	}
+
+	return nil
+}
+
+func CacheImages(images []string, cacheDir string) error {
+	var g errgroup.Group
+	for _, image := range images {
+		image := image
+		g.Go(func() error {
+			dst := filepath.Join(cacheDir, image)
+			dst = sanitizeCacheDir(dst)
+			if err := CacheImage(image, dst); err != nil {
+				return errors.Wrapf(err, "caching image %s", dst)
+			}
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		return errors.Wrap(err, "caching images")
+	}
+	glog.Infoln("Successfully cached all images.")
+	return nil
+}
+
+func LoadImages(cmd bootstrapper.CommandRunner, images []string, cacheDir string) error {
```

Comment:
This function looks almost exactly the same as the previous one. Any way you could refactor it to take the LoadFromCache/Cache functions as a parameter to avoid the repetition?

Reply:
Ah yeah good catch.  Will do that

Reply:
The only issue is that the signature is a little different - load needs a command runner to move the files into the VM/host, which cache doesn't.

Reply:
The functions have different signatures, so I've been having a little trouble figuring out how to share code here.  Any suggestions?

Reply:
I didn't fix this one.  I gave it an attempt, but all my attempts ended up way more verbose than the current solution.  

I think this would probably be easiest if Cache and Load were methods on the bootstrapper, so they could have the same implementation.  However, that limits when we can actually call them.  Since CacheImages doesn't actually require the bootstrapper, we can call it as soon as we start minikube up, rather than possibly waiting for SSH to be available.

Reply:
Sorry, one more quick idea here. Does it make sense to start loading each image as it becomes available in the cache? Right now you're downloading all the images in parallel, then loading all the images in parallel. What if it was "download and load" all images in parallel?

Reply:
I had that before, but I think this is actually better.  Combining cache and load forces us to wait for the machine to be SSH-able before we do anything (since we'll need a bootstrapper before we run load)

This caches in parallel as soon as we run a `minikube start`, but then as soon as the machine is ready, it tries a blocking load in parallel.  So both cache and load start as soon as they can.

Reply:
Ah, good call.